lib/User/UserEntry.php

```diff
@@ -0,0 +1,360 @@
+		$path = '';
+ 		$attr = $this->getAttributeName('homeFolderNamingRule', null);
+		if (is_string($attr) && strpos($attr, 'attr:') === 0 // TODO do faster startswith check
+			&& strlen($attr) > 5
+		) {
+			$attr = substr($attr, 5);
+
+			$path = $this->getAttributeValue($attr, '');
+		}
+
+		if ($path !== '') {
+			//if attribute's value is an absolute path take this, otherwise append it to data dir
+			//check for / at the beginning
+			if('/' !== $path[0]) {
+				$path = $this->config->getSystemValue('datadirectory',
+						\OC::$SERVERROOT.'/data' ) . '/' . $path;
+			}
+			return $path;
+		}
+
+		if(!is_null($attr)
+			&& $this->config->getAppValue('user_ldap', 'enforce_home_folder_naming_rule', true)
+		) {
+			// a naming rule attribute is defined, but it doesn't exist for that LDAP user
+			throw new \Exception('Home dir attribute can\'t be read from LDAP for uid: ' . $this->getUsername());
+		}
+
+		return null;
+	}
+
+	/**
+	 * @return string[]
+	 */
+	public function getMemberOfGroups() {
+		if (isset($this->ldapEntry['memberof'])) {
+			return $this->ldapEntry['memberof'];
+		}
+		return [];
+	}
+
+	/**
+	 * @brief reads the image from LDAP that shall be used as Avatar
+	 * @return string|null image binary data (provided by LDAP)
+	 */
+	public function getAvatarImage() {
+		$data = $this->getAttributeValue ('jpegPhoto', null, false);
+		if ($data === null) {
+			$data = $this->getAttributeValue ('thumbnailPhoto', null, false);
+		}
+		return $data;
+	}
+
+	/**
+	 * @return string[]
+	 */
+	public function getSearchTerms() {
+		$rawAttributes = $this->connection->ldapAttributesForUserSearch;
+		$attributes = empty($rawAttributes) ? [] : $rawAttributes;
+		// Get from LDAP if we don't have it already
+		$searchTerms = [];
+		foreach($attributes as $attr) {
+			$attr = strtolower($attr);
+			if (isset($this->ldapEntry[$attr])) {
+				foreach ($this->ldapEntry[$attr] as $value) {
+					$value = trim($value);
+					if (!empty($value)) {
+						$searchTerms[strtolower($value)] = true;
+					}
+				}
+			}
+		}
+
+		return array_keys($searchTerms);
+	}
+
+	/**
+	 * @param $configOption string
+	 * @param $default string
+	 * @return string
+	 */
+	private function getAttributeName($configOption, $default = '') {
+		$attributeName = trim(strtolower($this->connection->$configOption));
+
+		// strtolower() returns '' for null and false, which is what the connection initializes config options to
+		if ($attributeName === '') {
+			return $default;
+		}
+
+		return $attributeName;
+	}
+	/**
+	 * TODO allow passing in a verification function, eg for quota or uuid values
+	 * @param $attributeName string
+	 * @param $default string
+	 * @param $trim bool don't trim value, eg for binary data
+	 * @return string|null|mixed
+	 */
+	private function getAttributeValue($attributeName, $default = null, $trim = true) {
+		$attributeName = strtolower($attributeName); // all ldap keys are lowercase
+		if (isset($this->ldapEntry[$attributeName][0])) {
```

Comment:
Should we check for multivalued attributes? We should have a comment just in case.

Reply:
no. only search attributes may be multi value. updating phpdoc as well